File: Graphs/bfs_shortest_path.py
from get_adj_list import get_adj_list

def bfs_shortest_path(num_nodes, adj_list, start_node, end_node):
    """Find a shortest path from a start node to an end node."""

    # Initialize data structures
    path = []
    queue = [start_node]
    explored = [0] * (num_nodes + 1)
    parent = [0] * (num_nodes + 1)

    # Search until the queue is empty or the end node is found
    while queue != []:
        
        # Retrieve the next node from the queue
        node = queue.pop(0)

        # Mark the node as explored
        explored[node] = 1

        # Investigate node's neighbors.
        for neighbor in adj_list[node]:
                
            if explored[neighbor] == 0 and neighbor not in queue:

                # Mark the neighbor's parent.
                parent[neighbor] = node

                if neighbor == end_node:

                    # Re-trace the path from end node to start node
                    path.append(end_node)
                    while path[-1] != start_node:
                        path.append(parent[path[-1]])

                    return path[::-1]
                
                else:
                    queue.append(neighbor)

    return path


def main():

    # Define a graph
    num_nodes = 9
    edges = [
        [1,2],
        [2,3],
        [1,4],
        [2,6],
        [4,5],
        [5,6],
        [4,7],
        [5,7],
        [6,9],
        [8,9]
    ]

    # Get the adjacency list
    adj_list = get_adj_list(num_nodes, edges)

    # Return the BFS traversal for a starting node
    start_node = 6
    end_node = 7
    shortest_path = bfs_shortest_path(num_nodes, adj_list, start_node, end_node)
    print(shortest_path)
    

if __name__ == '__main__':
    main()

File: Cpp/bfs_shortest_path.cpp
#include <algorithm>
#include <iostream>
#include <queue>
#include <vector>
using namespace std;

vector<vector<int>> bfs(
    const vector<vector<int>> &grid,
    const vector<int> &start,
    const vector<int> &end
) {
    
    int num_rows = grid.size();
    int num_cols = grid[0].size();

    // Initialize the shortest path.
    vector<vector<int>> path;

    // Define the directions to travel.
    vector<vector<int>> directions = {
        {0, -1}, {0, 1}, {1, 0}, {-1, 0}
    };

    // Create a vector of explored nodes.
    vector<vector<int>> explored(
        num_rows, vector<int>(num_cols, 0)
    );

    // Create a vector of parent nodes.
    vector<vector<vector<int>>> parent(
        num_rows, vector<vector<int>>(num_cols, vector<int>(2))
    );

    // Create a queue of nodes to explore.
    queue<vector<int>> to_explore;

    // Add the start node to the queue and mark explored.
    to_explore.push(start);
    explored[start[0]][start[1]] = 1;

    // Search until the queue is empty. 
    while (!to_explore.empty()) {
        
        // Retrieve the next node from the queue and remove it.
        vector<int> node = to_explore.front();
        to_explore.pop();

        // Investigate node's neighbors. 
        for (vector<int> dir : directions) {

            vector<int> neighbor = {
                node[0] + dir[0],
                node[1] + dir[1]
            };

            // Return if end node is found.
            if (neighbor == end) {
                
                // Add the end node to the path.
                path.push_back(neighbor);

                // Add each node's parent to the path until reaching the start.
                while (node != start) {
                    path.push_back(node);
                    node = parent[node[0]][node[1]];
                }

                // Add the start node.
                path.push_back(node);

                // Reverse the path.
                reverse(path.begin(), path.end());

                return path;
            }


            if (
                // Check if the neighbor is in bounds.
                neighbor[0] >= 0 &&
                neighbor[1] >= 0 &&
                neighbor[0] < num_rows &&
                neighbor[1] < num_cols &&
                // Check if the neighbor is traversable.
                !grid[neighbor[0]][neighbor[1]] &&
                // Check if the neighbor is explored.
                !explored[neighbor[0]][neighbor[1]]
            ) {
                // Add the neighbor to the queue. 
                to_explore.push(neighbor);

                // Mark node as explored.
                explored[neighbor[0]][neighbor[1]] = 1;

                // Mark the neighbor's parent.
                parent[neighbor[0]][neighbor[1]] = node;
            }
        }
    }

    return path;
}


int main() {
    
    // Create a grid (0 traversable, 1 not).
    const vector<vector<int>> grid = {
        {0, 0, 1, 1, 0},
        {1, 0, 0, 1, 1},
        {0, 0, 0, 0, 1},
        {1, 1, 1, 0, 0},
        {0, 0, 0, 0, 1}
    };

    // Choose the start and end nodes.
    const vector<int> start = {0, 0};
    const vector<int> end = {4, 0};

    // Run BFS to find the shortest path.
    vector<vector<int>> path = bfs(grid, start, end);

    for (vector<int> node : path) {
        cout << "(" << node[0] << ", " << node[1] << ")" << endl;
    }

    return 0;
}